### PR TITLE
CAE

### PR DESCRIPTION
**Pull request by msJinLei**
## Description

* Implemented CAE by adding handler to http pipeline
* Improved error message when login is blocked by AAD
* Improved error message when silent reauthentication failed
* Enabled CAE for Get-AzTenant and Get-AzSubcription
* Added test cases

## Checklist

- [ ] I have read the [_Submitting Changes_](../blob/master/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/master/CONTRIBUTING.md)
- [ ] The title of the PR is clear and informative
- [ ] The appropriate `ChangeLog.md` file(s) has been updated:
    - For any service, the `ChangeLog.md` file can be found at `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`
    - A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header -- no new version header should be added
- [ ] The PR does not introduce [breaking changes](../blob/master/documentation/breaking-changes/breaking-changes-definition.md)
- [ ] If applicable, the changes made in the PR have proper test coverage
- [ ] For public API changes to cmdlets:
    - [ ] a cmdlet design review was approved for the changes in [this repository](https://github.com/Azure/azure-powershell-cmdlet-review-pr) (_Microsoft internal only_)
        - {Please put the link here}
    - [ ] the markdown help files have been regenerated using the commands listed [here](../blob/master/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)

### CODE REVIEW

File: tools/ScenarioTest.ResourceManager/Mocks/MockTokenAuthenticationFactory.cs [hunk 1]
```diff
@@ -121,7 +122,7 @@ public SubscriptionCloudCredentials GetSubscriptionCloudCredentials(IAzureContex
 
         public Microsoft.Rest.ServiceClientCredentials GetServiceClientCredentials(IAzureContext context)
         {
-            return new Microsoft.Rest.TokenCredentials(Token.AccessToken);
+            return new RenewingTokenCredential(Token);
```
Comment (msJinLei) [inline tools/ScenarioTest.ResourceManager/Mocks/MockTokenAuthenticationFactory.cs:125]: Revert

File: src/Aks/Aks.sln [hunk 2]
```diff
@@ -20,6 +20,10 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "ScenarioTest.ResourceManage
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "TestFx", "..\..\tools\TestFx\TestFx.csproj", "{BC80A1D0-FFA4-43D9-AA74-799F5CB54B58}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Azure.Core", "..\..\..\azure-sdk-for-net\sdk\core\Azure.Core\src\Azure.Core.csproj", "{773A5AC7-3CD2-4D80-84A0-01AAA8C79947}"
```
Comment (msJinLei) [inline src/Aks/Aks.sln:23]: Revert

File: src/Accounts/Authentication/Factories/ClientFactory.cs [hunk 3]
```diff
@@ -90,7 +100,9 @@ public ClientFactory()
                 types.Add(paramType);
                 if (paramType == typeof(DelegatingHandler[]))
                 {
-                    handlerList.AddRange(obj as DelegatingHandler[]);
+                    var handlers = obj as DelegatingHandler[];
+                    handlers = handlers.ToArray();
+                    handlerList.AddRange(handlers);
```
Comment (erich-wang) [inline src/Accounts/Authentication/Factories/ClientFactory.cs:105]: Is this change necessary?

File: src/Accounts/Authenticators/SilentAuthenticator.cs [hunk 4]
```diff
@@ -36,19 +38,21 @@ public override Task<IAccessToken> Authenticate(AuthenticationParameters paramet
...
                 EnableGuestTenantAuthentication = true,
                 ClientId = AuthenticationHelpers.PowerShellClientId,
                 Username = silentParameters.UserId,
                 AuthorityHost = new Uri(authority),
                 TenantId = tenantId,
             };
-
-            var cacheCredential = new SharedTokenCacheCredential(options);
+            azureCredentialFactory.Options = options;
+            var cacheCredential = azureCredentialFactory.CreateCredentialsFromOptions();
```
Comment (erich-wang) [inline src/Accounts/Authenticators/SilentAuthenticator.cs:53]: Is this change necessary?

**Review by erich-wang — CHANGES_REQUESTED** [review]

File: src/Accounts/Authentication/Utilities/Base64UrlHelpers.cs [hunk 5]
```diff
@@ -19,7 +19,7 @@
 namespace Microsoft.Azure.Commands.Profile.Utilities
 {
     //The source code is copied from https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/master/src/client/Microsoft.Identity.Client/Utils/Base64UrlHelpers.cs
-    internal static class Base64UrlHelpers
+    public static class Base64UrlHelpers
```
Comment (erich-wang) [inline src/Accounts/Authentication/Utilities/Base64UrlHelpers.cs:22]: use the one in common library

File: src/Accounts/Authenticators/MsalAccessToken.cs [hunk 6]
```diff
@@ -113,5 +114,16 @@ private bool IsNearExpiration()
             var timeUntilExpiration = ExpiresOn - DateTimeOffset.UtcNow;
             return timeUntilExpiration < ExpirationThreshold;
         }
+
+        public async ValueTask<bool> OnClaimsChallenageAsync(HttpRequestMessage request, string claimsChallenge, CancellationToken cancellationToken)
+        {
+            //TODO: Credential may not support
```
Comment (erich-wang) [inline src/Accounts/Authenticators/MsalAccessToken.cs:120]: check?

File: src/Accounts/Authenticators/SilentAuthenticator.cs [hunk 7]
```diff
@@ -36,19 +38,21 @@ public override Task<IAccessToken> Authenticate(AuthenticationParameters paramet
             var scopes = AuthenticationHelpers.GetScope(onPremise, resource);
             var authority = silentParameters.Environment.ActiveDirectoryAuthority;
 
-            AzureSession.Instance.TryGetComponent(nameof(PowerShellTokenCache), out PowerShellTokenCache tokenCache);
-            var options = new SharedTokenCacheCredentialOptions(tokenCache.TokenCache)
+            //AzureSession.Instance.TryGetComponent(nameof(PowerShellTokenCache), out PowerShellTokenCache tokenCache);
```
Comment (msJinLei) [inline src/Accounts/Authenticators/SilentAuthenticator.cs:41]: revert

File: tools/ScenarioTest.ResourceManager/Mocks/MockTokenAuthenticationFactory.cs [hunk 8]
```diff
@@ -133,12 +134,12 @@ public SubscriptionCloudCredentials GetSubscriptionCloudCredentials(IAzureContex
...
         public ServiceClientCredentials GetServiceClientCredentials(IAzureContext context, string targetEndpoint)
         {
-            return new Microsoft.Rest.TokenCredentials(Token.AccessToken);
+            return new RenewingTokenCredential(Token);
         }
 
         public Rest.ServiceClientCredentials GetServiceClientCredentials(string accessToken, Func<string> renew = null)
         {
-            throw new System.NotImplementedException();
+            return new RenewingTokenCredential(new ExternalAccessToken(accessToken));
```
Comment (msJinLei) [inline tools/ScenarioTest.ResourceManager/Mocks/MockTokenAuthenticationFactory.cs:142]: Check the change

**Comment by isra-fel** [comment]
Let me move the pull request to draft state. Please feel free to mark it "Ready for review" when you feel comfortable. Thanks

**Comment by msJinLei** [comment]
> Let me move the pull request to draft state. Please feel free to mark it "Ready for review" when you feel comfortable. Thanks

It is not because the feature is still in development that the PR is not merged. It is because the sever side is not ready that we cannot release the feature.

**Comment by isra-fel** [comment]
> It is not because the feature is still in development that the PR is not merged. It is because the sever side is not ready that we cannot release the feature.

Got it. It's just easier for maintanence. Thanks for understanding 😀

**Pull request merged**